robots.pm search_engines.pm and awstats_misc_tracker.js mods

visualperception opened the pull request:
a few minor robot search engine additions. Some regex corrections and cleanup and  
Added Device Pixel Ratio ( dpr ) to awstats_misc_tracker.js query string.
This has no effect on screen sizes report but does allow you to see in raw log files what the DPR of any device is. Works for current for IE (and edge I presume), webkit and mozilla.
awstats.pl screen size report should really be modified to include this but I'm no perl programmer.

visualperception committed:
Robots and Search Engines fixes and additions

A few more minor robots and search engines added. Some regex fixes and cleanups.

visualperception committed:
MInor regex fixes additions, tweaks and cleanup. 

MInor regex fixes additions, tweaks and cleanup.

visualperception committed:
Merge remote-tracking branch 'refs/remotes/eldy/develop' into develop

visualperception committed:
Revert "MInor regex fixes additions, tweaks and cleanup. "

This reverts commit 1baf3335fec4dfbed9f804c81361fc6347015d25.

visualperception committed:
Minor additions and a couple regex changes

Fixed a couple of my minor errors and added minor robots
also fixes my error of uploading files to wrong location

visualperception committed:
Added Device Pixel Ratio ( dpr ) to awstats_misc_tracker.js query string

modern monitors have varying device pixel ratios. Small screens such as IPhones use device pixel ratios ( dpr ) of higher than 1. Some have dpr of 2, some 3 and some as high as 5 or 6. But its not only mobile devices that use pixel doubling or trebling etc. The new 4K and 5K monitors use it as well.
Devices that use a dpr of not equal to 1 alter the window.screen width and height in the DOM according to the dpr. So an apple 5K monitor that has a true resolution of 5120x2880 pixels and uses a dpr of 2, returns window.screen width and height of 2560x1440.
It so happens that Apple produce a non retina monitor which has a dpr of 1 and window.screen of 2560x1440. 
Currently, the awstats screen sizes report will add both of these very different monitors to the same count which is technically incorrect.
I'm not perl programmer and don't know how to rectify the code to count correctly and output the DPR in the screen sizes report. However,

visualperception committed:
Add files via upload

eldy commented on wwwroot/cgi-bin/lib/robots.pm:
```diff
@@ -400,26 +404,26 @@
 'Yahoo![\x20]Slurp',
 'Baiduspider/',
-'Baiduspider\-image',
+'Baiduspider-image',
```
keys here are used into regex. Why removing the escaping char "\  ?

visualperception replied:
because the escaping is not necessary. The hyphen is only a special regex chararcter inside [ ]
Removing the escape where its not required lessens the load on regex and makes the files more readable.

visualperception replied:
Regarding the Device Pixel Ratio change, There are Apple monitors which have screen.width = 2560px and have a DPR of 1. And there are other Apple monitors which have screen.width = 2560 and have DPR of 2. They are very different monitors, the one with DPR =2 being a 5K monitor with twice the resolution. However the pixel doubling (DPR = 2 ) that the 5K monitor uses casues it to return a screen.width of 5120/DPR = 2560px to the DOM screen.width so currently we don't know if its a 5120 (5K) monitor or a 2560 monitor since they both use screen.width of 2560..
So now that there are many devices, especially mobile phones using a dpr of between 1 and upto 6 (so far I think) the screen sizes report doesn't give a true picture of what devices are being used.
I think people would be more interested in seeing what the dpr is which I realise is a significant change to the log output file and would require including screen dimension combined with dpr as the unique identifier which would make the screen size report longer as there would be more unique combinations than there are currently. The current screen size report is not accurate becasue of this since its not telling us which true monitor resolutions are being used. 
As an interim I just added the dpr into the query string so anyone who wants to check (me) can see it in the raw log files. It has no effect on reporting as it currently stands.

visualperception commented on wwwroot/cgi-bin/lib/robots.pm:
```diff
@@ -400,26 +404,26 @@
 'Yahoo![\x20]Slurp',
 'Baiduspider/',
-'Baiduspider\-image',
+'Baiduspider-image',
```
because the escaping is not necessary. The hyphen is only a special regex chararcter inside [ ]

visualperception replied:
because the escaping is not necessary. The hyphen is only a special regex chararcter inside [ ]

visualperception replied:
because the escaping is not necessary. The hyphen is only a special regex chararcter inside [ ]

visualperception replied:
because the escaping is not necessary. The hyphen is only a special regex chararcter inside square brackets

eldy commented on wwwroot/cgi-bin/lib/robots.pm:
```diff
@@ -2551,8 +2576,8 @@
-'bot[\s_+:,\.\;\/\\\-]','Unknown robot identified by bot\*',
-'[\s_+:,\.\;\/\\\-]bot','Unknown robot identified by \*bot',
+'bot[\s_+:,\.;/\\-]','Unknown robot identified by bot\*',
+'[\s_+:,\.;/\\-]bot','Unknown robot identified by \*bot',
```
Here the / is inside the [ ] , so we should keep the \/  instead of  /  to detect char /, shouldn't it ?

visualperception replied:
that's a forward slash which is not a special regex character so doesn't need escaping. i.e. without the escape \ then / is treated as / and not as a delimiter.
there may be confusion becasue in some implementations of regex / is used as a start end delimiter but not in this case I think. Infact in PHP which uses PCRE and requires the regex pattern to be enclosed in delimiters, / is often the chosen delimiter. But it is NOT being used as a delimiter in this case and is not a regex special character and so does not need escaping.
Also note when hyphen (-) is the last character inside character class [] it doesn't need to be escaped.  So you only need the \s for whitespace and to escape fullstop and the \ and hence all that is required is:
`'[\s_+:,\.;/\\-]bot'
`
I have been running it like this and it seems to be working fine.

visualperception replied:
that's a forward slash which is not a special regex character so doesn't need escaping. i.e. without the escape \ then / is treated as / and not as a delimiter.
there may be confusion becasue in some implementations of regex / is used as a start end delimiter but not in this case I think. Infact in PHP which uses PCRE and requires the regex pattern to be enclosed in delimiters, / is often the chosen delimiter. But it is being used as a delimiter in this case and is not a regex special character and so does not need escaping.
Also note when hyphen (-) is the last character inside character class [] it doesn't need to be escaped.  So you only need the \s for whitespace and to escape fullstop and the \ and hence all that is required is:
'[\s_+:,\.;/\\-]bot'

I have been running it like this and it seems to be working fine.

However, if Perl is using the PCRE functons for regex then maybe the escape would be required. I'm not 100% sure which regex implementation perl is using (not being a perl programmer).
I suppose that if there is any doubt then using an escape character removes any doubt even if it isn't really required.

visualperception commented:
I have looked into the / delimiter question and the result is it works fine in awstats without escaping / inside a character class. However I'm not 100 sure exactly why.
Firstly I think perl is using PCRE since PCRE is the perl implementation of regex.
I tried two online perl regex testers. The first one auto escaped the / which I had left out and the second one didn't. So one worked and one didn't which was inconclusive.
Further reading suggested that as with many special regex characters, / loses its meaning inside a character class.
I wrote a PHP test using preg_replace and that supposedly uses PCRE functions and it requires / to be escaped inside a character class.
So then I created a log record and added abc-bot to the UA and using my non escaped version in robots.txt I updated my stats. The result was that log record was identified as "Unknown robot identified by `\*`bot".
So awstats doesn't require the escape of / inside a character class []. But exactly why I don't know. Either the particular perl implementation of PCRE doesn't require escaping of / inside a character class OR the perl implementation of PCRE is auto escaping / inside a character class. I don't know which but either way it works fine as I suggested.
As always with regex, things can be somewaht unlear about what exactly is happening and why depending on the exact implementation of regex and in which prgramming language.

note: had the / in robots.txt acted as a delimiter, then regex should have failed with an error and what follows the / would not have been recognised to identify the log record as a bot. So it must be working fine unless detection regex is hardcoded elsewhere inside awstats.pl.

visualperception commented:
I'm running ActivePerl from activestate (on windows) . Other perl versions may differ so maybe the escape should be included.
